IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,471 @@
+                throttleDuration = (timeDiff > MAX_THROTTLING_TIME) ? MAX_THROTTLING_TIME : timeDiff;
+            }
+            break;
+        case MSIDThrottlingTypeUIRequired:
+            thumbprint = request.fullRequestThumbprint;
+            throttleTypeString = @"MSIDThrottlingTypeUIRequired";
+            throttleDuration = DefaultUIRequired;
+            break;
+        default:
+            break;
+    }
+    
+    MSIDThrottlingCacheRecord *record = [[MSIDThrottlingCacheRecord alloc] initWithErrorResponse:errorResponse
+                                                                                    throttleType:throttleTypeString
+                                                                                throttleDuration:throttleDuration];
+    [self.cacheService setObject:record forKey:thumbprint error:error];        
+}
+
+/**
+// TODO:  Huge TODO Here
+ */
+- (void)updateServerTelemetry:(MSIDThrottlingCacheRecord *)cacheRecord
+{
+    
+}
+
+- (NSDate *)getRetryDateFromErrorResponse:(NSError * _Nullable)errorResponse
+{
+    NSDate *retryHeaderDate = nil;
+    NSString *retryHeaderString = nil;
+    NSDictionary *headerFields = errorResponse.userInfo[MSIDHTTPHeadersKey];
+    retryHeaderString = headerFields[@"Retry-After"];
+    retryHeaderDate = [NSDate msidDateFromRetryHeader:retryHeaderString];
+    return retryHeaderDate;
+}
+
++ (BOOL)validateInput:(id<MSIDThumbprintCalculatable> _Nonnull)request
+{
+    return (!request.fullRequestThumbprint && !request.strictRequestThumbprint);
+}
+
+/**
+ Get last refresh time from our NSUserDefaults.
+ */
++ (NSDate *)getLastRefreshTimeWithContext:(id<MSIDRequestContext>)context
+                                    error:(NSError*__nullable*__nullable)error
+{
+    NSDictionary *searchQuery = [self prepareSearchQuery];
+    CFDictionaryRef result = nil;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)searchQuery, (CFTypeRef *)&result);
+    if (status != errSecSuccess)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Throttling: last refresh time not found with error code:%d", (int)status);
+        return nil;
+    }
+    
+    NSString *stringData = [(__bridge NSDictionary *)result objectForKey:(__bridge id)(kSecAttrService)];
+    
+    if (result)
+    {
+        CFRelease(result);
+    }
+    
+    if (!stringData || stringData.msidTrimmedString.length == 0)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDKeychainErrorDomain, status, @"Found empty keychain item.", nil, nil, nil, context.correlationId, nil, NO);
+        }
+        return nil;
+    }
+    
+    return [NSDate msidDateFromTimeStamp:stringData];
+}
+
++ (NSDictionary *)prepareSearchQuery
+{
+    static dispatch_once_t once;
+    static NSMutableDictionary *query = nil;
+    
+    dispatch_once(&once, ^{
+        query = [[NSMutableDictionary alloc] init];
+        NSString *keychainGroup = MSIDKeychainTokenCache.defaultKeychainGroup;
```

Comment:
This doesn't account for custom groups app can set, but see my comment above - perhaps we can eliminate keychain altogether to simplify.